awx/main/scheduler/kubernetes.py

```diff
@@ -0,0 +1,74 @@
+import time
+
+from django.conf import settings
+from kubernetes.client import Configuration as kubeconfig
+from kubernetes.client import CoreV1Api as v1_kube_api
+from kubernetes.client import ApiClient as kube_api_client
+
+from awx.main.utils.common import parse_yaml_or_json
+
+
+class PodManager(object):
+    credential = None
+    kube_api = None
+
+    def __init__(self, task):
+        self.task = task
+        self.credential = task.instance_group.credential
+        self.kube_api = self._kube_api_connection(self.credential)
+        self.pod_name = f"job-{task.id}"
+        self.pod_definition = self._pod_definition()
+
+    def deploy_pod(self):
+        if not self.credential.kubernetes:
+            raise RuntimeError
+
+        self.kube_api.create_namespaced_pod(body=self.pod_definition,
+                                            namespace='awx')
+
+        while True:
+            pod = self.kube_api.read_namespaced_pod(name=self.pod_name,
+                                                    namespace='awx')
+            if pod.status.phase != 'Pending':
+                break
+            time.sleep(1)
+
+        return pod
+
+    def _kube_api_connection(self, credential):
+        configuration = kubeconfig()
+        configuration.api_key["authorization"] = credential.get_input('bearer_token')
+        configuration.api_key_prefix['authorization'] = 'Bearer'
+        configuration.host = credential.get_input('host')
+        configuration.verify_ssl = False
+
+        return v1_kube_api(kube_api_client(configuration))
+
+    def _pod_definition(self):
+        default_pod_spec = {
+            "apiVersion": "v1",
+            "kind": "Pod",
+            "metadata": {
+                "name": self.pod_name
+            },
+            "spec": {
+                "containers": [{
+                    "name": self.pod_name,
+                    "image": settings.AWX_CONTAINER_GROUP_DEFAULT_IMAGE,
+                    "tty": True,
+                    "stdin": True,
+                    "imagePullPolicy": "Always",
+                    "command": [
+                        'sleep', 'infinity'
+                    ]
+                }]
+            }
+        }
+
+        if bool(self.task.instance_group.pod_spec_override):
+            pod_spec_override = parse_yaml_or_json(
+                self.task.instance_group.pod_spec_override)
+
+            return {**default_pod_spec, **pod_spec_override}
```

Comment:
This is going to be pretty cool - I'm wondering if we can somehow surface the pod spec defaults in the API somehow so that the UI can present it as the default value in the yaml editor.

Reply:
I am not opposed to the idea at all. Do you have any opinions about how you would like to see that surfaced?

Reply:
I don't have any particular opinions on how/where to surface these. That said, I think I've seen OPTIONS data used for sending over default values for fields. 

I checked our settings endpoints, which do something similar, and they provide default settings in the GET response _and_ in the OPTIONS data:

![Screenshot from 2019-08-12 12-47-51](https://user-images.githubusercontent.com/9753817/62882325-a3a9b000-bcff-11e9-98fb-c2b279c1de8a.png)

:man_shrugging: There may already be a convention for this, but I do not know it. From a UI perspective I think the most important thing would be that nothing needs to be saved or created through the API for the values to become available.

Reply:
Allowing user-provided pod-spec is risky. We would want more assertions around the end result of the pod spec so that we can proceed with certain assumptions in our code that references the podspec. i.e.

https://github.com/ansible/awx/blob/devel/awx/main/fields.py#L418 `json.Draft4Validator(schema).iter_errors` where schema looks like https://github.com/ansible/awx/blob/devel/awx/main/fields.py#L713